llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -8308,6 +8392,148 @@ llvm::Function *SPIRVToLLVM::createLibraryEntryFunc() {
+  constexpr ScratchBoundsCheckInfo nullValue{};
+  if (!getPipelineOptions()->enableScratchAccessBoundsChecks) {
+    return nullValue;
+  }
+
+  if (!toBeWrapped || toBeWrapped->getOpCode() != OpAccessChain) {
+    return nullValue;
+  }
+
+  SPIRVStorageClassKind pointerStorageClass = toBeWrapped->getType()->getPointerStorageClass();
+
+  // We don't want bounds checks if we are outside private or function storage class.
+  if (pointerStorageClass != StorageClassPrivate && pointerStorageClass != StorageClassFunction) {
+    return nullValue;
+  }
+
+  // Fetch the access chain behind the pointer the memory operation uses.
+  const auto accessChainBase = static_cast<SPIRVAccessChainBase *const>(toBeWrapped);
+  const auto accessChainIndices = accessChainBase->getIndices();
+  if (accessChainIndices.empty()) {
+    return nullValue;
+  }
+
+  SPIRVType *baseType = accessChainBase->getBase()->getType()->getPointerElementType();
+
+  if (!baseType) {
+    return nullValue;
+  }
+
+  // This will contain our final compare result.
+  Value *finalCmpResult = nullptr;
+
+  BasicBlock *entryBlock = getBuilder()->GetInsertBlock();
+  Function *parent = entryBlock->getParent();
+
+  // Build a chain of OOB checks by consuming all feasible access chain indices front to back.
+  for (SPIRVValue *currentIndex : accessChainIndices) {
+    // Handle either arrays or (nested) structs.
+    if (baseType->getOpCode() == OpTypeArray) {
+      if (const auto baseArray = static_cast<const SPIRVTypeArray *>(baseType)) {
+        // Add the bounds check.
+        const uint64_t arrayLength = baseArray->getArrayLength();
+        if (isConstantOpCode(currentIndex->getOpCode())) {
+          if (static_cast<SPIRVConstant *>(currentIndex)->getZExtIntValue() < arrayLength) {
+            baseType = baseType->getArrayElementType();
+            continue;
+          }
+        }
+
+        Value *comparisonValue = transValue(currentIndex, entryBlock->getParent(), entryBlock);
+        ConstantInt *arrayUpperBound = getBuilder()->getInt32(arrayLength);
+        auto cmpResult = getBuilder()->CreateICmpULT(comparisonValue, arrayUpperBound);
+
+        if (finalCmpResult) {
+          finalCmpResult = getBuilder()->CreateAnd(finalCmpResult, cmpResult);
+        } else {
+          finalCmpResult = cmpResult;
+        }
+
+        baseType = baseType->getArrayElementType();
+
+        if (!baseType) {
+          break;
+        }
+      }
+    } else if (baseType->getOpCode() == OpTypeStruct) {
+      // Fetch the struct member index.
+      assert(isConstantOpCode(currentIndex->getOpCode()));
+
+      const uint64_t structMemberIndex = static_cast<SPIRVConstant *>(currentIndex)->getZExtIntValue();
+      baseType = baseType->getStructMemberType(static_cast<size_t>(structMemberIndex));
+
+      if (!baseType) {
+        break;
+      }
+    } else {
+      break;
+    }
+
+    // Check if the next element is feasible for checking.
```

Comment:
Hi, you still need to add Null pointer check  here. there are still warnings

Reply:
Ran cppcheck locally. For me, there aren't any warnings anymore, hopefully this works.